core/embed/extmod/modtrezorio/modtrezorio-flash.h

```diff
@@ -118,3 +123,302 @@ STATIC const mp_obj_type_t mod_trezorio_FlashOTP_type = {
     .make_new = mod_trezorio_FlashOTP_make_new,
     .locals_dict = (void *)&mod_trezorio_FlashOTP_locals_dict,
 };
+
+/// class FlashArea:
+///     """
+///     Area of the flash memory
+///     """
+
+typedef struct _mp_obj_FlashArea_t {
+  mp_obj_base_t base;
+  const flash_area_t *area;
+} mp_obj_FlashArea_t;
+
+#define FLASH_READ_CHUNK_SIZE 1024
+#define PROGRESS_CHUNK_SIZE ((1024 / FLASH_READ_CHUNK_SIZE) * 16)
+
+static void ui_progress(mp_obj_t ui_wait_callback, uint32_t current) {
+  if (mp_obj_is_callable(ui_wait_callback)) {
+    mp_call_function_1_protected(ui_wait_callback, mp_obj_new_int(current));
+  }
+}
+
+/// def size(self) -> int:
+///     """
+///     Returns size of the flash area
+///     """
+STATIC mp_obj_t mod_trezorio_FlashArea_size(mp_obj_t obj_self) {
+  mp_obj_FlashArea_t *self = (mp_obj_FlashArea_t *)MP_OBJ_TO_PTR(obj_self);
+  return MP_OBJ_NEW_SMALL_INT(flash_area_get_size(self->area));
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mod_trezorio_FlashArea_size_obj,
+                                 mod_trezorio_FlashArea_size);
+
+/// def hash(
+///     self,
+///     offset: int,
+///     length: int,
+///     challenge: bytes | None = None,
+///     callback: Callable[[int], None] | None = None,
+/// ) -> bytes:
+///     """
+///     Computes a Blake2s hash of a segment of the flash area.
+///     Offset and length must be aligned to 1024 bytes.
+///     An optional challenge can be used as the Blake2s key.
+///     The progress callback will be invoked every 16 kB with the current
+///     position.
+///     """
+STATIC mp_obj_t mod_trezorio_FlashArea_hash(size_t n_args,
+                                            const mp_obj_t *args) {
+  mp_obj_FlashArea_t *self = (mp_obj_FlashArea_t *)MP_OBJ_TO_PTR(args[0]);
+  uint32_t offset = trezor_obj_get_uint(args[1]);
+  uint32_t length = trezor_obj_get_uint(args[2]);
+
+  if (offset % FLASH_READ_CHUNK_SIZE != 0 ||
+      length % FLASH_READ_CHUNK_SIZE != 0) {
+    mp_raise_ValueError("Offset and length must be aligned to 1024 bytes.");
+  }
+
+  mp_buffer_info_t challenge = {0};
+  if (n_args > 3 && args[3] != mp_const_none) {
+    mp_get_buffer_raise(args[3], &challenge, MP_BUFFER_READ);
+  }
+  mp_obj_t ui_wait_callback = mp_const_none;
+  if (n_args > 4) {
+    ui_wait_callback = args[4];
+  }
+
+  BLAKE2S_CTX ctx;
+
+  if (challenge.len != 0) {
+    if (blake2s_InitKey(&ctx, BLAKE2S_DIGEST_LENGTH, challenge.buf,
+                        challenge.len) != 0) {
+      mp_raise_msg(&mp_type_ValueError, "Invalid challenge.");
+    }
+  } else {
+    blake2s_Init(&ctx, BLAKE2S_DIGEST_LENGTH);
+  }
+
+  uint32_t area_size = flash_area_get_size(self->area);
+  if (area_size - offset < length) {
```

Comment:
Since we are dealing with unsigned values, I don't think this will detect `offset > area_size`, because the subtraction will wrap to a large value.